src/contracts/middleware/StakeRegistry.sol

```diff
@@ -0,0 +1,461 @@
+        //     }
+        // }     
+    }
+
+    // INTERNAL FUNCTIONS
+
+    function _getStakeUpdateIndexForOperatorIdForQuorumAtBlockNumber(bytes32 operatorId, uint8 quorumNumber, uint32 blockNumber) internal view returns(uint32) {
+        uint32 length = uint32(operatorIdToStakeHistory[operatorId][quorumNumber].length);
+        for (uint32 i = 0; i < length; i++) {
+            if (operatorIdToStakeHistory[operatorId][quorumNumber][length - i - 1].updateBlockNumber <= blockNumber) {
+                require(
+                    operatorIdToStakeHistory[operatorId][quorumNumber][length - i - 1].nextUpdateBlockNumber == 0 ||
+                    operatorIdToStakeHistory[operatorId][quorumNumber][length - i - 1].nextUpdateBlockNumber > blockNumber,
+                    "StakeRegistry._getStakeUpdateIndexForOperatorIdForQuorumAtBlockNumber: operatorId has no stake update at blockNumber"
+                );
+                return length - i - 1;
+            }
+        }
+        revert("StakeRegistry._getStakeUpdateIndexForOperatorIdForQuorumAtBlockNumber: no stake update found for operatorId and quorumNumber at block number");
+    }
+
+    function _setMinimumStakeForQuorum(uint8 quorumNumber, uint96 minimumStake) internal {
+        minimumStakeForQuorum[quorumNumber] = minimumStake;
+        emit MinimumStakeForQuorumUpdated(quorumNumber, minimumStake);
+    }
+
+    /** 
+     * @notice Updates the stake for the operator with `operatorId` for the specified `quorumNumbers`. The total stake
+     * for each quorum is updated accordingly in addition to the operator's individual stake history.
+     */ 
+    function _registerOperator(address operator, bytes32 operatorId, bytes memory quorumNumbers) internal {
+        // check the operator is registering for only valid quorums
+        require(uint8(quorumNumbers[quorumNumbers.length - 1]) < quorumCount, "StakeRegistry._registerOperator: greatest quorumNumber must be less than quorumCount");
+        OperatorStakeUpdate memory _newTotalStakeUpdate;
+        // add the `updateBlockNumber` info
+        _newTotalStakeUpdate.updateBlockNumber = uint32(block.number);
+        // for each quorum, evaluate stake and add to total stake
+        for (uint8 quorumNumbersIndex = 0; quorumNumbersIndex < quorumNumbers.length;) {
+            // get the next quorumNumber
+            uint8 quorumNumber = uint8(quorumNumbers[quorumNumbersIndex]);
+            // evaluate the stake for the operator
+            // since we don't use the first output, this will use 1 extra sload when deregistered operator's register again
+            (, uint96 stake) = _updateOperatorStake(operator, operatorId, quorumNumber);
+            // @JEFF: This reverts pretty late, but i think that's fine. wdyt?
+            // check if minimum requirement has been met, will be 0 if not
+            require(stake != 0, "StakeRegistry._registerOperator: Operator does not meet minimum stake requirement for quorum");
```

Comment:
seems hard to avoid a fairly late revert here; sometimes these things just are what they are. I think it's OK, but something to note when explaining some of the race conditions with this design.